src/GafferOSL/OSLObject.cpp

```diff
@@ -57,14 +61,31 @@ size_t OSLObject::g_firstPlugIndex;
 namespace
 {
 
-CompoundDataPtr prepareShadingPoints( const Primitive *primitive )
+
+CompoundDataPtr prepareShadingPoints( const Primitive *primitive, const std::string &requiredPrimvars )
 {
+	std::vector<std::string> primvarNames;
+	boost::algorithm::split( primvarNames, requiredPrimvars, boost::algorithm::is_any_of(" ") );
+
 	CompoundDataPtr shadingPoints = new CompoundData;
 	for( PrimitiveVariableMap::const_iterator it = primitive->variables.begin(), eIt = primitive->variables.end(); it != eIt; ++it )
 	{
-		// cast is ok - we're only using it to be able to reference the data from the shadingPoints,
-		// but nothing will modify the data itself.
-		shadingPoints->writable()[it->first] = boost::const_pointer_cast<Data>( it->second.data );
+		bool primvarUsed = std::find( primvarNames.begin(), primvarNames.end(), it->first ) != primvarNames.end();
+
+		// todo consider passing something like IndexedView to the ShadingEngine
+		// to avoid the expansion of indexed data.
+		// We only expand the primvar if it's actually used by the shader.
+		if ( it->second.indices && primvarUsed )
+		{
+			shadingPoints->writable()[it->first] = it->second.expandedData();
+		}
+		else
+		{
+			// cast is ok - we're only using it to be able to reference the data from the shadingPoints,
```

Comment:
We're adding the primvar here even if `primvarUsed` is false. That seems wrong, but if I understand correctly, is necessary because Constant primvars aren't listed in `requiredPrimvars`. In other words, `requiredPrimvars` is a misleading name - it's actually `resampledPrimVars`.

Reply:
Can we not also omit the non-indexed primvars if `needsAttribute()` is false?

Reply:
I though so also but I got a few test failures which I couldn't explain. Seems like we depend on all the privmars being present in the shadedPoint compound some how. I'll look into it further.

Reply:
Thanks - I feel like there's either a hole in my understanding here, or a bug somewhere, so it'd be good to get to the bottom of it...

Reply:
Looks like we need to have the "P" primvar to define the number of points to shade. Also I found problems with a unit test which read 'u'  so have forced globals to be required to be on the safe side.